rust/chains/hyperlane-cosmos/src/providers/grpc.rs

```diff
@@ -1,123 +1,241 @@
-    ) -> ChainResult<SimulateResponse>;
-
-    /// generate raw tx
-    async fn generate_raw_tx<I: IntoIterator<Item = cosmrs::Any> + Sync + Send>(
-        &self,
-        msgs: I,
-        gas_limit: Option<U256>,
-    ) -> ChainResult<Vec<u8>>;
-
-    /// send tx
+    /// Send a wasm tx.
     async fn wasm_send<T: Serialize + Sync + Send>(
         &self,
         payload: T,
         gas_limit: Option<U256>,
     ) -> ChainResult<TxResponse>;
 
-    /// simulate tx
-    async fn wasm_simulate<T: Serialize + Sync + Send>(
-        &self,
-        payload: T,
-    ) -> ChainResult<SimulateResponse>;
+    /// Estimate gas for a wasm tx.
+    async fn wasm_estimate_gas<T: Serialize + Sync + Send>(&self, payload: T) -> ChainResult<u64>;
 }
 
 #[derive(Debug)]
-/// Cosmwasm GRPC Provider
+/// CosmWasm GRPC provider.
 pub struct WasmGrpcProvider {
+    /// Connection configuration.
     conf: ConnectionConf,
-    address: H256,
+    /// A contract address that can be used as the default
+    /// for queries / sends / estimates.
+    contract_address: CosmosAddress,
+    /// Signer for transactions.
     signer: Option<Signer>,
+    /// GRPC Channel that can be cheaply cloned.
+    /// See https://docs.rs/tonic/latest/tonic/transport/struct.Channel.html#multiplexing-requests
+    channel: Channel,
 }
 
 impl WasmGrpcProvider {
-    /// create new Cosmwasm GRPC Provider
-    pub fn new(conf: ConnectionConf, locator: ContractLocator, signer: Option<Signer>) -> Self {
-        Self {
+    /// Create new CosmWasm GRPC Provider.
+    pub fn new(
+        conf: ConnectionConf,
+        locator: ContractLocator,
+        signer: Option<Signer>,
+    ) -> ChainResult<Self> {
+        let channel = Endpoint::new(conf.get_grpc_url())?.connect_lazy();
+        let contract_address = CosmosAddress::from_h256(locator.address, &conf.get_prefix())?;
+
+        Ok(Self {
             conf,
-            address: locator.address,
+            contract_address,
             signer,
-        }
+            channel,
+        })
+    }
+
+    /// Gets a signer, or returns an error if one is not available.
+    fn get_signer(&self) -> ChainResult<&Signer> {
+        self.signer
+            .as_ref()
+            .ok_or(ChainCommunicationError::SignerUnavailable)
+    }
+}
+
+impl WasmGrpcProvider {
+    /// Generates an unsigned SignDoc for a transaction.
+    async fn generate_unsigned_sign_doc(
+        &self,
+        msgs: Vec<cosmrs::Any>,
+        gas_limit: u64,
+    ) -> ChainResult<SignDoc> {
+        // As this function is only used for estimating gas or sending transactions,
+        // we can reasonably expect to have a signer.
+        let signer = self.get_signer()?;
+        let account_info = self.account_query(signer.address.clone()).await?;
+        let current_height = self.latest_block_height().await?;
+        let timeout_height = current_height + TIMEOUT_BLOCKS;
+
+        let tx_body = tx::Body::new(
+            msgs,
+            "",
```

Comment:
cleaner to use `Default::default()` here

Reply:
the expected value is `impl Into<String>` so I'll just do String::default()